File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/modules/grades/GradeDao.java
package pl.szczodrzynski.edziennik.data.db.modules.grades;

import android.util.LongSparseArray;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.RawQuery;
import androidx.room.Transaction;
import androidx.sqlite.db.SimpleSQLiteQuery;
import androidx.sqlite.db.SupportSQLiteQuery;

import java.util.List;

import static pl.szczodrzynski.edziennik.data.db.modules.metadata.Metadata.TYPE_GRADE;

@Dao
public abstract class GradeDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract long add(Grade grade);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void addAll(List<Grade> gradeList);

    @Query("DELETE FROM grades WHERE profileId = :profileId")
    public abstract void clear(int profileId);

    @Query("DELETE FROM grades WHERE profileId = :profileId AND gradeType = :type")
    public abstract void clearWithType(int profileId, int type);

    @Query("DELETE FROM grades WHERE profileId = :profileId AND gradeSemester = :semester")
    public abstract void clearForSemester(int profileId, int semester);

    @Query("DELETE FROM grades WHERE profileId = :profileId AND gradeSemester = :semester AND gradeType = :type")
    public abstract void clearForSemesterWithType(int profileId, int semester, int type);

    @RawQuery(observedEntities = {Grade.class})
    abstract LiveData<List<GradeFull>> getAll(SupportSQLiteQuery query);
    public LiveData<List<GradeFull>> getAll(int profileId, String filter, String orderBy) {
        return getAll(new SimpleSQLiteQuery("SELECT \n" +
                "*, \n" +
                "teachers.teacherName || ' ' || teachers.teacherSurname AS teacherFullName\n" +
                "FROM grades \n" +
                "LEFT JOIN subjects USING(profileId, subjectId)\n" +
                "LEFT JOIN teachers USING(profileId, teacherId)\n" +
                "LEFT JOIN metadata ON gradeId = thingId AND thingType = " + TYPE_GRADE + " AND metadata.profileId = "+profileId+"\n" +
                "WHERE grades.profileId = "+profileId+" AND "+filter+"\n" +
                "ORDER BY "+orderBy)); // TODO: 2019-04-30 why did I add sorting by gradeType???
    }
    public LiveData<List<GradeFull>> getAllOrderBy(int profileId, String orderBy) {
        return getAll(profileId, "1", orderBy);
    }
    public LiveData<List<GradeFull>> getAllWhere(int profileId, String filter) {
        return getAll(profileId, filter, "addedDate DESC");
    }

    @RawQuery
    abstract List<GradeFull> getAllNow(SupportSQLiteQuery query);
    public List<GradeFull> getAllNow(int profileId, String filter) {
        return getAllNow(new SimpleSQLiteQuery("SELECT \n" +
                "*, \n" +
                "teachers.teacherName || ' ' || teachers.teacherSurname AS teacherFullName\n" +
                "FROM grades \n" +
                "LEFT JOIN subjects USING(profileId, subjectId)\n" +
                "LEFT JOIN teachers USING(profileId, teacherId)\n" +
                "LEFT JOIN metadata ON gradeId = thingId AND thingType = " + TYPE_GRADE + " AND metadata.profileId = "+profileId+"\n" +
                "WHERE grades.profileId = "+profileId+" AND "+filter+"\n" +
                "ORDER BY addedDate DESC"));
    }
    public List<GradeFull> getNotNotifiedNow(int profileId) {
        return getAllNow(profileId, "notified = 0");
    }
    public List<GradeFull> getAllWithParentIdNow(int profileId, long parentId) {
        return getAllNow(profileId, "gradeParentId = "+parentId);
    }

    @RawQuery
    abstract GradeFull getNow(SupportSQLiteQuery query);
    public GradeFull getNow(int profileId, String filter) {
        return getNow(new SimpleSQLiteQuery("SELECT \n" +
                "*, \n" +
                "teachers.teacherName || ' ' || teachers.teacherSurname AS teacherFullName\n" +
                "FROM grades \n" +
                "LEFT JOIN subjects USING(profileId, subjectId)\n" +
                "LEFT JOIN teachers USING(profileId, teacherId)\n" +
                "LEFT JOIN metadata ON gradeId = thingId AND thingType = " + TYPE_GRADE + " AND metadata.profileId = "+profileId+"\n" +
                "WHERE grades.profileId = "+profileId+" AND "+filter+"\n" +
                "ORDER BY addedDate DESC"));
    }
    public GradeFull getByIdNow(int profileId, long gradeId) {
        return getNow(profileId, "gradeId = "+gradeId);
    }

    @Query("UPDATE grades SET gradeClassAverage = :classAverage, gradeColor = :color WHERE profileId = :profileId AND gradeId = :gradeId")
    public abstract void updateDetailsById(int profileId, long gradeId, float classAverage, int color);

    @Query("UPDATE metadata SET addedDate = :addedDate WHERE profileId = :profileId AND thingType = "+TYPE_GRADE+" AND thingId = :gradeId")
    public abstract void updateAddedDateById(int profileId, long gradeId, long addedDate);

    @Transaction
    public void updateDetails(int profileId, LongSparseArray<Float> gradeAverages, LongSparseArray<Long> gradeAddedDates, LongSparseArray<Integer> gradeColors) {
        for (int i = 0; i < gradeAverages.size(); i++) {
            long gradeId = gradeAverages.keyAt(i);
            float classAverage = gradeAverages.valueAt(i);
            long addedDate = gradeAddedDates.valueAt(i);
            int color = gradeColors.valueAt(i);
            updateDetailsById(profileId, gradeId, classAverage, color);
            updateAddedDateById(profileId, gradeId, addedDate);
        }
    }

    @Query("SELECT gradeId FROM grades WHERE profileId = :profileId ORDER BY gradeId")
    public abstract List<Integer> getIds(int profileId);
    @Query("SELECT gradeClassAverage FROM grades WHERE profileId = :profileId ORDER BY gradeId")
    public abstract List<Float> getClassAverages(int profileId);
    @Query("SELECT gradeColor FROM grades WHERE profileId = :profileId ORDER BY gradeId")
    public abstract List<Integer> getColors(int profileId);
    @Query("SELECT addedDate FROM metadata WHERE profileId = :profileId AND thingType = "+TYPE_GRADE+" ORDER BY thingId")
    public abstract List<Long> getAddedDates(int profileId);
    @Transaction
    public void getDetails(int profileId, LongSparseArray<Long> gradeAddedDates, LongSparseArray<Float> gradeAverages, LongSparseArray<Integer> gradeColors) {
        List<Integer> ids = getIds(profileId);
        List<Float> classAverages = getClassAverages(profileId);
        List<Integer> colors = getColors(profileId);
        List<Long> addedDates = getAddedDates(profileId);
        for (int index = 0; index < ids.size(); index++) {
            if (classAverages.size() > index) {
                gradeAverages.put(ids.get(index), classAverages.get(index));
            }
            if (colors.size() > index) {
                gradeColors.put(ids.get(index), colors.get(index));
            }
            if (addedDates.size() > index) {
                gradeAddedDates.put(ids.get(index), addedDates.get(index));
            }
        }
    }

    public LiveData<List<GradeFull>> getAllFromDate(int profileId, int semester, long date) {
        return getAllWhere(profileId, "gradeSemester = " + semester + "  AND addedDate > " + date);
    }
}

File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/dao/GradeDao.kt
/*
 * Copyright (c) Kuba Szczodrzyński 2020-4-24.
 */
package pl.szczodrzynski.edziennik.data.db.dao

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Query
import androidx.room.RawQuery
import androidx.room.Transaction
import androidx.sqlite.db.SupportSQLiteQuery
import pl.szczodrzynski.edziennik.App
import pl.szczodrzynski.edziennik.annotation.SelectiveDao
import pl.szczodrzynski.edziennik.annotation.UpdateSelective
import pl.szczodrzynski.edziennik.data.db.AppDb
import pl.szczodrzynski.edziennik.data.db.entity.Grade
import pl.szczodrzynski.edziennik.data.db.entity.Metadata
import pl.szczodrzynski.edziennik.data.db.full.GradeFull
import pl.szczodrzynski.edziennik.utils.models.Date
import java.util.*
import kotlin.collections.component1
import kotlin.collections.component2
import kotlin.collections.set

@Dao
@SelectiveDao(db = AppDb::class)
abstract class GradeDao : BaseDao<Grade, GradeFull> {
    companion object {
        private const val QUERY = """
            SELECT 
            *, 
            teachers.teacherName ||" "|| teachers.teacherSurname AS teacherName
            FROM grades
            LEFT JOIN teachers USING(profileId, teacherId)
            LEFT JOIN subjects USING(profileId, subjectId)
            LEFT JOIN metadata ON gradeId = thingId AND thingType = ${Metadata.TYPE_GRADE} AND metadata.profileId = grades.profileId
        """

        private const val ORDER_BY = """ORDER BY addedDate DESC"""
    }

    private val selective by lazy { GradeDaoSelective(App.db) }

    @RawQuery(observedEntities = [Grade::class])
    abstract override fun getRaw(query: SupportSQLiteQuery): LiveData<List<GradeFull>>
    @RawQuery(observedEntities = [Grade::class])
    abstract override fun getOne(query: SupportSQLiteQuery): LiveData<GradeFull?>

    // SELECTIVE UPDATE
    @UpdateSelective(primaryKeys = ["profileId", "gradeId"], skippedColumns = ["addedDate", "gradeClassAverage"])
    override fun update(item: Grade) = selective.update(item)
    override fun updateAll(items: List<Grade>) = selective.updateAll(items)

    // CLEAR
    @Query("DELETE FROM grades WHERE profileId = :profileId")
    abstract override fun clear(profileId: Int)
    // REMOVE NOT KEPT
    @Query("DELETE FROM grades WHERE keep = 0")
    abstract override fun removeNotKept()

    // GET ALL - LIVE DATA
    fun getAll(profileId: Int) =
            getRaw("$QUERY WHERE grades.profileId = $profileId $ORDER_BY")
    fun getAllFromDate(profileId: Int, date: Date) =
            getRaw("$QUERY WHERE grades.profileId = $profileId AND addedDate > ${date.inMillis} $ORDER_BY")
    fun getAllBySubject(profileId: Int, subjectId: Long) =
            getRaw("$QUERY WHERE grades.profileId = $profileId AND subjectId = $subjectId $ORDER_BY")
    fun getAllOrderBy(profileId: Int, orderBy: String) =
            getRaw("$QUERY WHERE grades.profileId = $profileId ORDER BY $orderBy")

    // GET ALL - NOW
    fun getAllNow(profileId: Int) =
            getRawNow("$QUERY WHERE grades.profileId = $profileId $ORDER_BY")
    fun getNotNotifiedNow() =
            getRawNow("$QUERY WHERE notified = 0 $ORDER_BY")
    fun getNotNotifiedNow(profileId: Int) =
            getRawNow("$QUERY WHERE grades.profileId = $profileId AND notified = 0 $ORDER_BY")
    fun getByParentIdNow(profileId: Int, parentId: Long) =
            getRawNow("$QUERY WHERE grades.profileId = $profileId AND gradeParentId = $parentId $ORDER_BY")

    // GET ONE - NOW
    fun getByIdNow(profileId: Int, id: Long) =
            getOneNow("$QUERY WHERE grades.profileId = $profileId AND gradeId = $id")

    @Query("UPDATE grades SET keep = 0 WHERE profileId = :profileId AND gradeType = :type")
    abstract fun dontKeepWithType(profileId: Int, type: Int)

    @Query("UPDATE grades SET keep = 0 WHERE profileId = :profileId AND gradeSemester = :semester")
    abstract fun dontKeepForSemester(profileId: Int, semester: Int)

    @Query("UPDATE grades SET keep = 0 WHERE profileId = :profileId AND gradeSemester = :semester AND gradeType = :type")
    abstract fun dontKeepForSemesterWithType(profileId: Int, semester: Int, type: Int)



    // GRADE DETAILS - MOBIDZIENNIK
    @Query("UPDATE grades SET gradeClassAverage = :classAverage, gradeColor = :color WHERE profileId = :profileId AND gradeId = :gradeId")
    abstract fun updateDetailsById(profileId: Int, gradeId: Long, classAverage: Float, color: Int)

    @Query("UPDATE grades SET addedDate = :addedDate WHERE profileId = :profileId AND gradeId = :gradeId")
    abstract fun updateAddedDateById(profileId: Int, gradeId: Long, addedDate: Long)

    @Transaction
    open fun updateDetails(profileId: Int, gradeAverages: SortedMap<Long, Float>, gradeAddedDates: SortedMap<Long, Long>, gradeColors: SortedMap<Long, Int>) {
        for ((gradeId, addedDate) in gradeAddedDates) {
            val classAverage = gradeAverages[gradeId] ?: 0.0f
            val color = gradeColors[gradeId] ?: 0xff2196f3.toInt()
            updateDetailsById(profileId, gradeId, classAverage, color)
            updateAddedDateById(profileId, gradeId, addedDate)
        }
    }

    @Query("SELECT gradeId FROM grades WHERE profileId = :profileId ORDER BY gradeId")
    abstract fun getIds(profileId: Int): List<Long>

    @Query("SELECT gradeClassAverage FROM grades WHERE profileId = :profileId ORDER BY gradeId")
    abstract fun getClassAverages(profileId: Int): List<Float>

    @Query("SELECT gradeColor FROM grades WHERE profileId = :profileId ORDER BY gradeId")
    abstract fun getColors(profileId: Int): List<Int>

    @Query("SELECT addedDate FROM grades WHERE profileId = :profileId ORDER BY gradeId")
    abstract fun getAddedDates(profileId: Int): List<Long>

    @Transaction
    open fun getDetails(profileId: Int, gradeAddedDates: SortedMap<Long, Long>, gradeAverages: SortedMap<Long, Float>, gradeColors: SortedMap<Long, Int>) {
        val ids = getIds(profileId).iterator()
        val classAverages = getClassAverages(profileId).iterator()
        val colors = getColors(profileId).iterator()
        val addedDates = getAddedDates(profileId).iterator()
        while (ids.hasNext() && classAverages.hasNext() && colors.hasNext() && addedDates.hasNext()) {
            val gradeId = ids.next()
            gradeAverages[gradeId] = classAverages.next()
            gradeColors[gradeId] = colors.next()
            gradeAddedDates[gradeId] = addedDates.next()
        }
    }
}
